datafusion-examples/examples/parquet_encryption_with_kms.rs

```diff
@@ -0,0 +1,205 @@
+
+    let _ = ctx
+        .register_listing_table(
+            "parquet_table",
+            &table_path,
+            listing_options.clone(),
+            None,
+            None,
+        )
+        .await?;
+
+    let df = ctx.sql("SELECT * FROM parquet_table").await?;
+    let plan = df.create_physical_plan().await?;
+
+    let mut batch_stream = execute_stream(plan.clone(), ctx.task_ctx())?;
+    println!("Reading encrypted Parquet as a RecordBatch stream");
+    while let Some(batch) = batch_stream.next().await {
+        let batch = batch?;
+        println!("Read batch with {} rows", batch.num_rows());
+    }
+    println!("Finished reading");
+    Ok(())
+}
+
+// Custom options used to configure our encryption factory
+extensions_options! {
+   pub struct EncryptionConfig {
+       pub footer_key_id: Option<String>, default = None
+   }
+}
+
+impl ConfigExtension for EncryptionConfig {
+    const PREFIX: &'static str = "kms_encryption";
+}
+
+/// Wrapper type for `CryptoFactory` to allow implementing the `EncryptionFactory` trait
+struct KmsEncryptionFactory {
+    crypto_factory: CryptoFactory,
+}
+
+/// `EncryptionFactory` is a trait defined by DataFusion that allows generating
+/// file encryption and decryption properties.
+/// The `FileSinkConfig is provided so that the schema may be used to dynamically specify
+/// per-column encryption.
+impl EncryptionFactory for KmsEncryptionFactory {
+    fn get_file_encryption_properties(
+        &self,
+        options: &TableParquetOptions,
+        sink_config: &FileSinkConfig,
+    ) -> Result<FileEncryptionProperties> {
+        let config: &EncryptionConfig = options.extensions.get().ok_or_else(|| {
+            DataFusionError::Configuration(
+                "EncryptionConfig options not registered".to_owned(),
+            )
+        })?;
+        let footer_key_id = config.footer_key_id.as_ref().cloned().ok_or_else(|| {
+            DataFusionError::Configuration("Footer key id not configured".to_owned())
+        })?;
+        // We could configure per-column keys using the provided schema,
+        // but for simplicity this example uses uniform encryption.
+        let config =
+            EncryptionConfiguration::builder(footer_key_id.to_owned()).build()?;
+        // Similarly, KMS connection could be configured from the options.
+        let kms_config = Arc::new(KmsConnectionConfig::default());
+        Ok(self
+            .crypto_factory
+            .file_encryption_properties(kms_config, &config)?)
+    }
+
+    fn get_file_decryption_properties(
+        &self,
+        options: &TableParquetOptions,
+        file_path: &str,
+    ) -> Result<FileDecryptionProperties> {
+        let config = DecryptionConfiguration::builder().build()?;
+        let kms_config = Arc::new(KmsConnectionConfig::default());
+        Ok(self
+            .crypto_factory
+            .file_decryption_properties(kms_config, config)?)
+    }
+}
```

Comment:
The factory part looks good here. The key will be integrating it into ParquetOpener but I think we can do that in one place.

Reply:
Yeah I'm not sure how easy it will be to get access to the encryption factory when we need to use it to get file decryption properties. It looks like maybe we can do that when we create the `ParquetSource` in [`ParquetFormat::create_physical_plan`](https://github.com/apache/datafusion/blob/efd9587b2f271dd54c31e19416f04810f42675ca/datafusion/datasource-parquet/src/file_format.rs#L409).